File: 03.CPP/1749.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>

class Solution {
public:
    int maxAbosoluteSum(std::vector<int>& nums) {
        int maxPosSum = nums[0];
        int maxNegSum = nums[0];
        int maxAbsSum = std::abs(nums[0]);

        for (int i = 1; i < nums.size(); i++) {
            maxPosSum = std::max(nums[i], nums[i] + maxPosSum);
            maxNegSum = std::min(nums[i], nums[i] + maxNegSum);
            
            maxAbsSum = std::max(
                { maxAbsSum, std::abs(maxPosSum), std::abs(maxNegSum) }
            );
        }

        return maxAbsSum;
    }
};

int main() {
    Solution sol;

    std::vector<int> nums = {1, -3, 2, 3, -4};
    int result = sol.maxAbsoluteSum(nums);

    std::cout << result << "\n";
    return 0;
}

File: 01.Python/1749.py
from typing import List

class Solution:
    def maxAbsoluteSum(self, nums: List[int]) -> int:
        posLocalSum = nums[0]
        negLocalSum = nums[0]
        absMaxSum = 0

        for i in range(1, len(nums)):
            posLocalSum = max(nums[i], nums[i] + posLocalSum)
            negLocalSum = min(nums[i], nums[i] + negLocalSum)

            absMaxSum = max(absMaxSum, abs(posLocalSum), abs(negLocalSum))

        return absMaxSum

sol = Solution()

nums = [1, -3, 2, 3, -4]
result = sol.maxAbsoluteSum(nums)

print(result)